### PR TITLE
[ONNX FE] Support value freezing without specifying type

### PR DESCRIPTION
**Pull request by rkazants**
**Details:** It will allow to freeze Parameter inputs without specifying its type. Implement get_element_type method for ONNX InputModel. Implement tests with freezing integer input and float models.

**Tickets:** 98484, 98149

Signed-off-by: Kazantsev, Roman <roman.kazantsev@intel.com>

### CODE REVIEW

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 1]
```diff
@@ -193,6 +193,27 @@ void InputModel::set_element_type(const ov::frontend::Place::Ptr& place, const n
...
 }
 
+ov::element::Type InputModel::get_element_type(const ov::frontend::Place::Ptr& place) const {
+    std::string tensor_name;  // name of the model input which should be reshaped
+    const auto input_edge = std::dynamic_pointer_cast<PlaceInputEdge>(place);
+    const auto output_edge = std::dynamic_pointer_cast<PlaceOutputEdge>(place);
+    if (input_edge) {
+        const auto tensor_names = input_edge->get_source_tensor()->get_names();
+        OPENVINO_ASSERT(!tensor_names.empty(),
+                        "Cannot retrieve source tensor name for this InputEdge and thus partial shape.");
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.cpp:203]:
```suggestion
                        "Cannot retrieve source tensor name for this InputEdge and thus its element type.");
```

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 2]
```diff
@@ -193,6 +193,27 @@ void InputModel::set_element_type(const ov::frontend::Place::Ptr& place, const n
...
+        tensor_name = tensor_names[0];
+    } else if (output_edge) {
+        const auto tensor_names = output_edge->get_target_tensor()->get_names();
+        OPENVINO_ASSERT(!tensor_names.empty(),
+                        "Cannot retrieve target tensor name for this OutputEdge and thus partial shape.");
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.cpp:208]:
```suggestion
                        "Cannot retrieve target tensor name for this OutputEdge and thus its element type.");
```

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 3]
```diff
@@ -193,6 +193,27 @@ void InputModel::set_element_type(const ov::frontend::Place::Ptr& place, const n
...
+ov::element::Type InputModel::get_element_type(const ov::frontend::Place::Ptr& place) const {
+    std::string tensor_name;  // name of the model input which should be reshaped
+    const auto input_edge = std::dynamic_pointer_cast<PlaceInputEdge>(place);
+    const auto output_edge = std::dynamic_pointer_cast<PlaceOutputEdge>(place);
+    if (input_edge) {
+        const auto tensor_names = input_edge->get_source_tensor()->get_names();
+        OPENVINO_ASSERT(!tensor_names.empty(),
+                        "Cannot retrieve source tensor name for this InputEdge and thus partial shape.");
+        tensor_name = tensor_names[0];
+    } else if (output_edge) {
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.cpp:205]: From the parameter's freezing perspective the output edge handling is not required. This code can still stay here for potential future use.

**Review by tomdol — CHANGES_REQUESTED** [review]

File: src/frontends/onnx/frontend/src/input_model.cpp (hunk 1 above)
Comment (rkazants) [reply src/frontends/onnx/frontend/src/input_model.cpp]: Great catch! Fix it.

File: src/frontends/onnx/frontend/src/input_model.cpp (hunk 3 above)
Comment (rkazants) [reply src/frontends/onnx/frontend/src/input_model.cpp:205]: Let leave it here then

**Review by tomdol — APPROVED** [review]

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 4]
```diff
@@ -193,6 +193,27 @@ void InputModel::set_element_type(const ov::frontend::Place::Ptr& place, const n
     m_editor->set_input_types(m);
 }
 
+ov::element::Type InputModel::get_element_type(const ov::frontend::Place::Ptr& place) const {
+    std::string tensor_name;  // name of the model input which should be reshaped
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/input_model.cpp:197]: nit: The comment about reshape seems not related to the `tensor_name` usage in `get_element_type`
```suggestion
    std::string tensor_name;
```

**Review by mitruska — APPROVED** [review]

**Review by mbencer — CHANGES_REQUESTED** [review]
Please add the test to `openvino/src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py` (it can be done separetelly if the change has to be merge now). It can look like:
```
def test_get_element_type():
    skip_if_onnx_frontend_is_disabled()
    fe = fem.load_by_framework(framework=ONNX_FRONTEND_NAME)
    model = fe.load("input_model_2.onnx")

    in1 = model.get_place_by_tensor_name(tensor_name="in1")
    assert model.get_element_type(in1) == Type.f32 # works correctly
    
    in1_input_edge = in1.get_consuming_ports()[0]
    assert model.get_element_type(in1_input_edge) == Type.f32 # works correctly
    
    #in1_output_edge = in1.get_producing_port() # in1 is an input of the model and doesn't have producing port exception
    #assert model.get_element_type(in1_output_edge) == Type.f32
```
What's more I can't see scenario where `PlaceOutputEdge` can work now.
As I undestand parameter of the model (`TensorPlace`) can't be mapped 1:1 to `OutputEdge`. Please look at the image below:
image
For getting shape from internal part of a graph we are using information propagated via `ONNX_NAMESPACE::shape_inference::InferShapes` (provided by onnx). There is also sth like `TypeAndShapeInferenceFunction`, but it looks like overkill for the current purposes.
So as sum-up, `PlaceOutputEdge` path can be removed in my opinion (ideally assert with a message)

**Comment by rkazants** [comment]
> Please add the test to `openvino/src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py` (it can be done separetelly if the change has to be merge now). It can look like:
> 
> ```
> def test_get_element_type():
>     skip_if_onnx_frontend_is_disabled()
>     fe = fem.load_by_framework(framework=ONNX_FRONTEND_NAME)
>     model = fe.load("input_model_2.onnx")
> 
>     in1 = model.get_place_by_tensor_name(tensor_name="in1")
>     assert model.get_element_type(in1) == Type.f32 # works correctly
>     
>     in1_input_edge = in1.get_consuming_ports()[0]
>     assert model.get_element_type(in1_input_edge) == Type.f32 # works correctly
>     
>     #in1_output_edge = in1.get_producing_port() # in1 is an input of the model and doesn't have producing port exception
>     #assert model.get_element_type(in1_output_edge) == Type.f32
> ```
> 
> What's more I can't see scenario where `PlaceOutputEdge` can work now. As I undestand parameter of the model (`TensorPlace`) can't be mapped 1:1 to `OutputEdge`. Please look at the image below: image For getting shape from internal part of a graph we are using information propagated via `ONNX_NAMESPACE::shape_inference::InferShapes` (provided by onnx). There is also sth like `TypeAndShapeInferenceFunction`, but it looks like overkill for the current purposes. So as sum-up, `PlaceOutputEdge` path can be removed in my opinion (ideally assert with a message)

We should stay handling output edges because Parameter nodes (for which we pass place name) 

> Please add the test to `openvino/src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py` (it can be done separetelly if the change has to be merge now). It can look like:
> 
> ```
> def test_get_element_type():
>     skip_if_onnx_frontend_is_disabled()
>     fe = fem.load_by_framework(framework=ONNX_FRONTEND_NAME)
>     model = fe.load("input_model_2.onnx")
> 
>     in1 = model.get_place_by_tensor_name(tensor_name="in1")
>     assert model.get_element_type(in1) == Type.f32 # works correctly
>     
>     in1_input_edge = in1.get_consuming_ports()[0]
>     assert model.get_element_type(in1_input_edge) == Type.f32 # works correctly
>     
>     #in1_output_edge = in1.get_producing_port() # in1 is an input of the model and doesn't have producing port exception
>     #assert model.get_element_type(in1_output_edge) == Type.f32
> ```
> 
> What's more I can't see scenario where `PlaceOutputEdge` can work now. As I undestand parameter of the model (`TensorPlace`) can't be mapped 1:1 to `OutputEdge`. Please look at the image below: image For getting shape from internal part of a graph we are using information propagated via `ONNX_NAMESPACE::shape_inference::InferShapes` (provided by onnx). There is also sth like `TypeAndShapeInferenceFunction`, but it looks like overkill for the current purposes. So as sum-up, `PlaceOutputEdge` path can be removed in my opinion (ideally assert with a message)

Discussed offline. The implementation is optimal now. I will just to add one more test case to get element type for the Constant node.

**Review by mbencer — APPROVED** [review]
LGTM, thanks
